### PR TITLE
FIX: handle more thread pool edge cases

### PR DESCRIPTION
**Pull request by SamSaffron**
1. When there were zero threads, it could take a very long time to process first job
2. Properly scale up thread queue depending on actual busy work
3. Align APIs with concurrent

### CODE REVIEW

File: lib/scheduler/thread_pool.rb [hunk 1]
```diff
@@ -8,15 +8,19 @@ module Scheduler
...
 
   class ThreadPool
     class ShutdownError < StandardError
     end
 
-    def initialize(min_threads:, max_threads:, idle_time:)
+    def initialize(min_threads:, max_threads:, idle_time: nil)
+      # this is there just for cases where people want to use this as
+      # a fixed size thread pool
+      idle_time ||= 30
```
Comment (tgxworld) [inline lib/scheduler/thread_pool.rb:23]: Is there something significant about the number `30` as I'm wondering why setting it to 30 would mean it become a fixed size thread pool

File: lib/scheduler/thread_pool.rb [hunk 2]
```diff
@@ -45,37 +50,45 @@ def post(&block)
...
-      deadline = Process.clock_gettime(Process::CLOCK_MONOTONIC) + timeout
-      threads_to_join.each do |thread|
-        remaining_time = deadline - Process.clock_gettime(Process::CLOCK_MONOTONIC)
-        break if remaining_time <= 0
-        if !thread.join(remaining_time)
-          Rails.logger.error "ThreadPool: Failed to join thread within timeout\n#{thread.backtrace.join("\n")}"
-          failed_to_shutdown = true
+        if failed_to_shutdown
+          @mutex.synchronize { @threads.each(&:kill) }
+          raise ShutdownError, "Failed to shutdown ThreadPool within timeout" if failed_to_shutdown
```
Comment (tgxworld) [inline lib/scheduler/thread_pool.rb:80]:
```suggestion
          raise ShutdownError, "Failed to shutdown ThreadPool within timeout"
```

This is already within the `if failed_to_shutdown` conditional.

File: lib/scheduler/thread_pool.rb [hunk 3]
```diff
@@ -115,24 +129,40 @@ def thread_loop
...
               work = nil
               done = true
             end
           end
 
+          @busy_threads << Thread.current if work
+
+          if !done && work && @queue.length > 0 && @threads.length < @max_threads &&
+               @busy_threads.length == @threads.length
+            spawn_thread
```
Comment (tgxworld) [inline lib/scheduler/thread_pool.rb:153]: The parent of this thread will be that of another "worker" thread and I'm not completely sure how that might affection exceptions within a thread. Would it be safer and simpler to spin up this thread in `post` instead?

File: spec/lib/scheduler/thread_pool_spec.rb [hunk 4]
```diff
@@ -165,6 +174,24 @@
...
+
+      done = Queue.new
+      pool.post { done << :done }
+
+      # should happen in less than 1 second
+      Timeout.timeout(1) { expect(done.pop).to eq(:done) }
+
+      pool.shutdown
+      pool.wait_for_termination
+      expect(pool.stats[:thread_count]).to eq(0)
```
Comment (tgxworld) [inline spec/lib/scheduler/thread_pool_spec.rb:191]: I'm not sure if we need these lines cause it is testing `shutdown` logic and don't seem to be related to the test description.

File: lib/scheduler/thread_pool.rb (hunk 1 above)
Comment (SamSaffron) [reply lib/scheduler/thread_pool.rb:23]: will add to the comment, basically in a "fixed" mode stuff still idles per the current design, so this is a default.

File: lib/scheduler/thread_pool.rb (hunk 2 above)
Comment (SamSaffron) [reply lib/scheduler/thread_pool.rb:80]: oops thanks

File: lib/scheduler/thread_pool.rb (hunk 3 above)
Comment (SamSaffron) [reply lib/scheduler/thread_pool.rb:153]: too complicated to get the mechanics right, we would not be spinning up threads in time or have a giant lock on post that heavily slows down posting.

File: spec/lib/scheduler/thread_pool_spec.rb (hunk 4 above)
Comment (SamSaffron) [reply spec/lib/scheduler/thread_pool_spec.rb:191]: happy to amend the description.

**Review by tgxworld — APPROVED** [review]
